python/shotgun_menus/entity_field_menu.py

```diff
@@ -0,0 +1,311 @@
+        return action
+
+    def __del__(self):
+        """
+        Destructor
+        """
+        if self._owns_task_manager:
+            shotgun_globals.unregister_bg_task_manager(self._task_manager)
+
+    def _on_about_to_show(self):
+        """
+        Lazy load the menu.  This is because it is possible to have have cycles when traversing
+        through the possible bubbled fields, so it is impossible to build the entire nested menu.
+        """
+        if not self._initialized:
+            # need to wait until there is a schema available before populating the menu
+            shotgun_globals.run_on_schema_loaded(self._populate)
+            self._initialized = True
+
+    def _populate(self):
+        """
+        Build the menu
+        """
+        field_infos = []
+        bubble_fields = {}
+
+        # gather needed field info
+        for field in shotgun_globals.get_entity_fields(self._sg_entity_type):
+            # convert field to bubbled form
+            bubbled_field = self._get_bubbled_name(field)
+
+            # apply any needed filtering
+            if self._field_filter is not None and not self._field_filter(bubbled_field):
+                continue
+
+            # grab display names
+            display_name = shotgun_globals.get_field_display_name(self._sg_entity_type, field)
+            field_infos.append({"field": field, "name": display_name, "bubbled": bubbled_field})
+
+            # grab info to build bubbled menu
+            try:
+                # grab the entity types this field can bubble to
+                valid_types = shotgun_globals.get_valid_types(self._sg_entity_type, field)
+                valid_type_names = [shotgun_globals.get_type_display_name(valid_type) for valid_type in valid_types]
+
+                if valid_types:
+                    bubble_fields[field] = {
+                        "valid_types": valid_types,
+                        "valid_type_names": valid_type_names,
+                        "name": display_name,
+                    }
+            except Exception:
+                # not a field that can be bubbled
+                pass
+
+        # sort by display name
+        field_infos.sort(key=lambda item: item["name"])
+
+        # add in all fields other than audit fields
+        audit_fields = []
+        for field_info in field_infos:
+            if field_info["field"] in self._AUDIT_FIELDS:
+                audit_fields.append(field_info)
+            else:
+                self._add_qaction(field_info["bubbled"], field_info["name"])
+
+        # now the audit fields
+        if audit_fields:
+            self.addSeparator()
+            self.add_label("Audit Fields")
+            for field_info in audit_fields:
+                self._add_qaction(field_info["field"], field_info["name"])
+
+        # and finally bubble fields
+        if bubble_fields:
+            self.addSeparator()
+            self.add_label("Linked Fields")
```

Comment:
Would be pretty nice if we could get the entity type icons from the shotgun globals to decorate the menu items for the linked fields! :)